api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt

```diff
@@ -0,0 +1,245 @@
+            ),
+            arrayOf(
+                arrayOf(
+                    floatArrayOf(1.0f, 3.0f, 1.0f),
+                    floatArrayOf(6.0f, -2.5f, 4.0f),
+                    floatArrayOf(7.0f, 0.0f, 5.0f),
+                    floatArrayOf(1.0f, 2.0f, 4.0f)
+                ),
+                arrayOf(
+                    floatArrayOf(7.0f, -3.0f, 2.0f),
+                    floatArrayOf(1.0f, 2.0f, 2.0f),
+                    floatArrayOf(3.0f, 5.0f, -2.0f),
+                    floatArrayOf(3.0f, -1.0f, 0.0f)
+                ),
+            ),
+        ),
+    )
+
+    private val inputShape: Shape = Shape.make(
+        input.size.toLong(),
+        input[0].size.toLong(),
+        input[0][0].size.toLong(),
+        input[0][0][0].size.toLong(),
+        input[0][0][0][0].size.toLong(),
+    )
+
+    @Test
+    fun default() {
+        val layer = AvgPool3D()
+        val expected = arrayOf(
+            arrayOf(
+                arrayOf(
+                    arrayOf(
+                        floatArrayOf(18.0f/8, 4.5f/8, 16.0f/8),
+                        floatArrayOf(9.5f/8, 12.0f/8, 13.5f/8),
+                    ),
+                ),
+            ),
+        )
+
+        EagerSession.create().use {
+            val tf = Ops.create()
+            layer.build(tf, KGraph(Graph().toGraphDef()), inputShape)
+
+            val inputOp = tf.constant(input)
+            val isTraining = tf.constant(true)
+            val numberOfLosses =  tf.constant(1.0f)
+            val output = layer.forward(tf, inputOp, isTraining, numberOfLosses).asOutput()
+
+            // Check output shape is correct.
+            val expectedShape = intArrayOf(input.size, 1, 1, 2, input[0][0][0][0].size)
+            Assertions.assertArrayEquals(
+                expectedShape,
+                output.shape().toIntArray()
+            )
+
+            // Check output values are correct.
+            val actual = Array(input.size) {
+                Array(1) { Array(1) { Array(2) { FloatArray(input[0][0][0][0].size) } } }
+            }
+            output.tensor().copyTo(actual)
+            for (i in expected.indices) {
+                for (j in expected[i].indices) {
+                    for (k in expected[i][j].indices) {
+                        for (l in expected[i][j][k].indices) {
+                            Assertions.assertArrayEquals(
+                                expected[i][j][k][l],
+                                actual[i][j][k][l],
+                                EPS
+                            )
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun withPaddingAndStride() {
+        val layer = AvgPool3D(strides = longArrayOf(1, 1, 1, 1, 1), padding = ConvPadding.SAME)
+        val expected = arrayOf(
+            arrayOf(
+                arrayOf(
+                    arrayOf(
+                        floatArrayOf(18.0f/8, 4.5f/8, 16.0f/8),
```

Comment:
did you take these values from Keras tests?

Reply:
No, I created the test data myself :)

Reply:
Not bad:)
Sometimes I just copied test data from Keras; sometimes, if test data seems not appropriate, I use Keras like a test machine writing some simple test in Python to put some test data in inputs and get outputs. A few times, I ignored this method and just used tested primitive itself to get data, but it has a bug in implementation and, as a result, a wrong test:(( But I agree that test inputs/outputs for pooling could be calculated manually (just shared my experience)

Reply:
Yeah, actually I have created test data for all test cases I have written so far; but certainly it's not a sustainable approach and using another tested tool, like Keras or maybe even numpy, for generating test data seems to be a better and more reliable approach.